qiskit_ibm_runtime/transpiler/passes/basis/fold_rzz_angle.py

```diff
@@ -0,0 +1,239 @@
+    with angle of arbitrary real numbers.
+
+    .. note::
+        This pass doesn't transform the circuit when the
+        Rzz gate angle is an unbound parameter.
+        In this case, the user must assign a gate angle before
+        transpilation, or be responsible for choosing parameters
+        from the calibrated range of [0, pi/2].
+    """
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        self._run_inner(dag)
+        return dag
+
+    def _run_inner(self, dag: DAGCircuit) -> bool:
+        """Mutate the input dag to fix non-ISA Rzz angles."""
+        modified = False
+        for node in dag.op_nodes():
+            if isinstance(node.op, ControlFlowOp):
+                modified_blocks = False
+                new_blocks = []
+                for block in node.op.blocks:
+                    block_dag = circuit_to_dag(block)
+                    if self._run_inner(block_dag):
+                        # Store circuit with Rzz modification
+                        new_blocks.append(dag_to_circuit(block_dag))
+                        modified_blocks = True
+                    else:
+                        # Keep original circuit to save memory
+                        new_blocks.append(block)
+                if modified_blocks:
+                    dag.substitute_node(
+                        node,
+                        node.op.replace_blocks(new_blocks),
+                        inplace=True,
+                    )
+                    modified = True
+                continue
+
+            if not isinstance(node.op, RZZGate):
+                continue
+
+            angle = node.op.params[0]
+            if isinstance(angle, ParameterExpression) or 0 <= angle <= pi / 2:
+                # Angle is an unbound parameter or a calibrated value.
+                continue
+
+            # Modify circuit around Rzz gate to address non-ISA angles.
+            modified = True
+            wrap_angle = np.angle(np.exp(1j * angle))
+            if 0 <= wrap_angle <= pi / 2:
+                # In the first quadrant after phase wrapping.
+                # We just need to remove 2pi offset.
+                dag.substitute_node(
+                    node,
+                    RZZGate(wrap_angle),
+                    inplace=True,
+                )
+                continue
+            if pi / 2 < wrap_angle <= pi:
+                # In the second quadrant.
+                replace = _quad2(wrap_angle, node.qargs)
+            elif -pi <= wrap_angle <= -pi / 2:
+                # In the third quadrant.
+                replace = _quad3(wrap_angle, node.qargs)
+            elif -pi / 2 < wrap_angle < 0:
+                # In the forth quadrant.
+                replace = _quad4(wrap_angle, node.qargs)
+            else:
+                raise RuntimeError("Unreacheable.")
+            dag.substitute_node_with_dag(node, replace)
+        return modified
+
+
+def _quad2(angle: float, qubits: Tuple[Qubit, ...]) -> DAGCircuit:
```

Comment:
I think the `_quad` functions should be static functions inside the `FoldRzzAngle` class, @wshanks do you agree?

Reply:
Done in 4b08e7e0